pkg/reconciler/v1alpha1/taskrun/resources/input_resource_test.go

```diff
@@ -0,0 +1,365 @@
+							ResourceRef: v1alpha1.PipelineResourceRef{
+								Name: "workspace",
+							},
+						},
+					},
+				},
+			},
+		},
+		build:   build,
+		wantErr: false,
+		want: &buildv1alpha1.Build{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Build",
+				APIVersion: "build.knative.dev/v1alpha1"},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "build-from-repo",
+				Namespace: "marshmallow",
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion:         "pipeline.knative.dev/v1alpha1",
+						Kind:               "TaskRun",
+						Name:               "build-from-repo-run",
+						Controller:         &boolTrue,
+						BlockOwnerDeletion: &boolTrue,
+					},
+				},
+			},
+			Spec: buildv1alpha1.BuildSpec{
+				Source: &buildv1alpha1.SourceSpec{
+					Git: &buildv1alpha1.GitSourceSpec{
+						Url:      "https://github.com/grafeas/kritis",
+						Revision: "master",
+					},
+				},
+			},
+		},
+	}, {
+		desc: "set service account if provided",
+		task: &v1alpha1.Task{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "build-from-repo",
+				Namespace: "marshmallow",
+			},
+			Spec: v1alpha1.TaskSpec{
+				Inputs: &v1alpha1.Inputs{
+					Sources: []v1alpha1.Source{
+						v1alpha1.Source{
+							Name: "workspace-sa",
+							Type: "git",
+						},
+					},
+				},
+			},
+		},
+		taskRun: taskRun,
+		build:   build,
+		wantErr: false,
+		want: &buildv1alpha1.Build{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Build",
+				APIVersion: "build.knative.dev/v1alpha1"},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "build-from-repo",
+				Namespace: "marshmallow",
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion:         "pipeline.knative.dev/v1alpha1",
+						Kind:               "TaskRun",
+						Name:               "build-from-repo-run",
+						Controller:         &boolTrue,
+						BlockOwnerDeletion: &boolTrue,
+					},
+				},
+			},
+			Spec: buildv1alpha1.BuildSpec{
+				Source: &buildv1alpha1.SourceSpec{
+					Git: &buildv1alpha1.GitSourceSpec{
+						Url:      "https://github.com/grafeas/kritis",
+						Revision: "master",
+					},
+				},
+				ServiceAccountName: "kritis-service-account",
+			},
+		},
+	}, {
+		desc: "invalid resource name",
+		task: &v1alpha1.Task{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "build-from-repo",
+				Namespace: "marshmallow",
+			},
+			Spec: v1alpha1.TaskSpec{
+				Inputs: &v1alpha1.Inputs{
+					Sources: []v1alpha1.Source{
+						v1alpha1.Source{
+							Name: "workspace-invalid",
+							Type: "git",
+						},
+					},
+				},
+			},
+		},
+		taskRun: taskRun,
+		build:   build,
+		wantErr: true,
+		want:    nil,
+	},
+	} {
+		setUp()
+		t.Run(c.desc, func(t *testing.T) {
+			got, err := AddInputResource(c.build, c.task, c.taskRun, pipelineResourceLister, logger)
+			if (err != nil) != c.wantErr {
+				t.Errorf("Test: %q; NewControllerConfigFromConfigMap() error = %v, WantErr %v", c.desc, err, c.wantErr)
+			}
```

Comment:
im kind of making this comment everywhere, but what are your thoughts about having separate test functions for expected failures and successes? I feel like mixing them and having to handle both here clutters up the test a bit

Reply:
I'm okay with splitting them if you prefer that, big table tests are hard to read for sure